Feature/3266 apply error pages

alexis-ld opened the pull request:
**Component**: ui, errorpages

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**:  #3266 
We want to add error pages to metalK8S.

**Summary**:
This adds :
* a 404 page to undefined routes.
* a 500 page when the config file is not found/valid
* a 500 page on Navbar fail
* an Auth error page when the user is not authenticated




---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #3266 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

bert-e commented:
# Hello alexis-ld,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality

ChengYanJin commented on ui/cypress/integration/error-pages.spec.js:
```diff
@@ -0,0 +1,100 @@
+    cy.stubHistory();
+
+    cy.get('[data-cy="sc-error-page404"]').should('have.length', 1);
+  });
```
IMO, can be removed since it's the same case as the next one.

ChengYanJin commented on ui/src/components/Navbar.js:
```diff
@@ -187,19 +188,9 @@ function useLogoutEffect(
 export function Navbar() {
   return (
-    <ErrorBoundary FallbackComponent={ErrorFallback}>
+    <ErrorBoundary fallbackRender={() => <Redirect to='/500'/>}>
```
We should directly render the `ErrorPage` instead of having route `/500` and redirecting to it.

ChengYanJin commented on ui/src/containers/Layout.js:
```diff
@@ -160,6 +167,25 @@ const Layout = () => {
+                      locale={language}
+                    />
+                  )}
+                />
```
Should be removed

ChengYanJin commented on ui/cypress/integration/error-pages.spec.js:
```diff
@@ -0,0 +1,75 @@
+    cy.stubHistory();
+
+    cy.get('[data-cy="sc-error-page500"]').should('have.length', 1);
+  });
```
Should be removed since we don't need to have `/500` route

ChengYanJin commented on ui/cypress/integration/error-pages.spec.js:
```diff
@@ -0,0 +1,75 @@
+    cy.visit('/undefined-route-that-doesnt-exists');
+    cy.stubHistory();
+
+    cy.get('[data-cy="sc-error-page404"]').should('have.length', 1);
```
IMO, we should test if the text on the 404 page render correctly because that's what the users see and really matter. 

Same for all the error pages

alexis-ld replied:
@ChengYanJin do you mean testing if the 404 text is rendered properly in there ?

I think that may be a bad idea because this text is directly into core-ui ErrorPage component, meaning if we update the text in core-ui it would make the metal tests fail.

ChengYanJin replied:
I think we don't need to check the actual text, but add an `aria-label` for the text element in core-ui

ChengYanJin commented on ui/cypress/integration/error-pages.spec.js:
```diff
@@ -0,0 +1,75 @@
+    cy.get('[data-cy="sc-error-page500"]').should('have.length', 1);
+  });
+
+  it('500 error page should contains support link if set to config.json', () => {
+    let url_support = 'https://mysupportlink';
+    cy.fixture('config.json').then((config) => {
+      config['url_support'] = url_support;
+      cy.route2('GET', '/config.json', config);
+    });
+
+    cy.visit('/500');
+    cy.stubHistory();
+
+    cy.get('[data-cy="sc-error-page500"]').should('have.length', 1);
+    cy.get('.sc-error-page500 * a')
+      .should('have.attr', 'href')
+      .and('include', url_support);
+  });
+});
```
For me, this should not be an independent scenario to test, should be an assertion in a scenario. For example, when we test the navbar failing scenario, we can check if the support link is rendered on the 500 error page.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin

ChengYanJin commented on ui/cypress/fixtures/config.json:
```diff
@@ -1,12 +1,13 @@
+  "url_alerts": "/shell/alerts.1.0.0.js",
+  "alerts_lib_version": "1.0.0",
+  "url_navbar_config": "/shell/config.json",
+  "url_support": "https://support.scality.com/hc/en-us"
```
don't see why we need a real link here. 
And also please add an empty entry in metalk8s UI `config.json`
`"url_support": ""`

gdemonet replied:
Actually, if we could have something like `https://github.com/scality/metalk8s/discussions/new`, that'd be awesome! (not in the test, in the default config used when deploying MetalK8s)

gdemonet commented on ui/public/config.json:
```diff
@@ -10,5 +10,5 @@
   "url_navbar_config": "/shell/config.json",
   "alerts_lib_version": "1.0.0",
-  "url_support": ""
+  "url_support": "https://github.com/scality/metalk8s/discussions/new"
```
Thanks! :+1:

alexis-ld commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin







*The following options are set:* **approve**

ChengYanJin committed:
shell-ui: Render the error page if the navbar fail to load

Refs: #3266

ChengYanJin committed:
ui: Add error auth page if not authenticate in PrivateRoute

Refs: #3266

ChengYanJin committed:
ui: Add error page when fail to load navbar or alertprovider

Refs: #3266

ChengYanJin committed:
ui/layout: Add 404 page for unregistered route

Add redirect from `/` to `/nodes`
The homepage should be the dashboard page once it's ready

Refs: #3266

ChengYanJin committed:
ui: Add url_support in config.json

Refs: #3266

ChengYanJin committed:
ui/tests: Add integration test for the error page

Refs: #3266

ChengYanJin committed:
ui: Add error page when we can't load the config

bert-e commented:
# Conflict




There is a conflict between your branch `feature/3266-apply-error-pages` and the
destination branch `development/2.9`.

Please resolve the conflict on **the feature branch** (`feature/3266-apply-error-pages`).

```bash
 $ git fetch
 $ git checkout origin/feature/3266-apply-error-pages
 $ git merge origin/development/2.9
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:feature/3266-apply-error-pages
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin







*The following options are set:* **approve**

JBWatenbergScality committed:
ui: Fix a race condition leading to an infinte authentication loop on firefox

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality










*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/c7794100a9d46e3476085a1bf88a0a275dc84e74) did not succeed in branch feature/3266-apply-error-pages.




*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.9`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.9`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`



Please check the status of the associated issue None.

Goodbye alexis-ld.
